### ISSUE TITLE
cross-platform, cross-compiler builds

### ISSUE THREAD

**Issue by ndarilek** [synthizer/synthizer#1]
Getting the following when attempting to build this in VSCode. Happens from the console too, but I haven't used Cmake in a while, thought I may be getting the incantations wrong, and figured I'd trust VSCode to get that right:

```
Cannot find source file:

  third_party/wdl/wdl/audiobuffercontainer.cpp

Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
.hpp .hxx .in .txx
```

Then below:

```
{
	"resource": "/home/nolan/src/synthizer/cmake/wdl.txt",
	"owner": "cmake-configure-diags",
	"severity": 8,
	"message": "No SOURCES given to target: wdl",
	"source": "CMake (add_library)",
	"startLineNumber": 20,
	"startColumn": 1,
	"endLineNumber": 20,
	"endColumn": 10000,
	"relatedInformation": [
		{
			"startLineNumber": 15,
			"startColumn": 1,
			"endLineNumber": 15,
			"endColumn": 1000,
			"message": "In call to 'include' here",
			"resource": "/home/nolan/src/synthizer/CMakeLists.txt"
		}
	]
}
```

Looks like you're missing some vendored source?

As an aside, would you be interested in a GitHub Action that just runs a build to catch issues like these? If so, I'm happy to contribute one. I know this project is fairly early, so a full CI setup is overkill, but I find that even a simple barebones setup is useful even when starting out.

Thanks for building Synthizer.

**Comment by camlorn** [comment]
VSCode doesn't work because VSCode's C++ support is beta and I managed to break it very badly somewhere along the way by combining Gnu and MSVC extensions, and then it turned out that for some reason the CMake support can't parse errors on my machine.  I still use it as the IDE, but always build from the command prompt.  You probably want:

```
mkdir build
cd build
cmake -G Ninja -DCMAKE_C_COMPILER="c:/program files/llvm/bin" -DCMAKE_CXX_COMPILER="c:/program files/llvm/bin"
ninja
```

You'll need VS 2019 and Clang 9 or later installed (possibly Clang 10).

Now that there's a C API (or the beginnings of one) and I have a 3 day weekend, I'm hoping to get at least Appveyor going with artifacts, and also proper build instructions.

**Comment by camlorn** [comment]
Sorry, correction.  It's:

```
cmake -G Ninja -DCMAKE_C_COMPILER="c:/program files/llvm/bin/clang.exe" -DCMAKE_CXX_COMPILER="c:/program files/llvm/bin/clang.exe"
```

You have to point it at the compiler itself, not the directories.  Sorry, I'm typing these from memory, as after the first build it's just ninja in the build directory.

**Comment by Keithcat1** [comment]
I don't think Clang 10 works.
Last I tried, I got this.
My build commands are as you specified. I did custom-build my version of clang though.
v

**Comment by ndarilek** [comment]
Thanks for the build commands.


As far as this particular issue goes, I've submitted #2. I'm attempting 
to build this on Linux (understanding it almost certainly won't work) 
and you had thirdparty/wdl/WDL as a directory, but thirdparty/wdl/wdl in 
your CMake configuration. Now I can at least generate Makefiles, which 
is where I leave this for now.


Thanks for the specific build commands. I'll submit another PR adding 
them to the README.md if it isn't done in a few days, and once I have a 
bit more time.

**Comment by camlorn** [comment]
Semi-officially making this the "I tried to build it and the world exploded" issue.

@ndarilek 
Yeah, good catch.  Windows directories are case insensitive.

This is probably 95% of the way to building on Linux.  I don't know what the last 5% is and the TBD Python bindings will probably take us further away from it, but in general I'll accept contributions for this and will try not to break it.  At the moment try not to break it means there will be CI--probably Travis--and I will consider it a problem if CI starts failing.

@Keithcat1 
The missing things are x86 intrinsics, and I don't maintain miniaudio.  Are you on a non-X86 platform?

If you are, I would expect that Clang would error at compile time instead of link time, though.  Miniaudio is supposed to be detecting that properly.

You say you've built clang yourself.  Is it possible that you've somehow broken it?  Mine is:

```
(base) PS C:\Program Files\LLVM\bin> .\clang.exe -v
clang version 10.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
```

Which I am literally building with right now, so it does work.

I'd suggest getting your clang from releases.llvm.org unless you have a good reason to be doing otherwise.

**Comment by camlorn** [comment]
Reopening because we're tracking more than just Linux path case sensitivity.

**Comment by Keithcat1** [comment]
@camlorn 
I'm on Windows 64-bit and building Clang with LTO and maybe PGO if I can get it working.
How can I include the missing features in the Clang build?
It works properly if I use the official Clang 10 build.

**Comment by camlorn** [comment]
I don't have many answers for you, as I've never built Clang.  I do know that debug symbols are working on my machine and in CI, and I have edited CMakeCache.txt to include lto in the past without issue.  The trick there is to also make clang use llvm's linker instead.  To my knowledge, LTO is included in Clang 10, and PGO might be also.  I haven't done this in long enough that I don't know the options offhand.

For the Python bindings, you're probably not going to get LTO out of it.  Python compiles via MSVC and links to the Clang library, and there's no way to change this that I know of.

I wouldn't expect much benefit from either LTO or PGO as applied to Synthizer itself, which is why I'm not showing a lot of interest.  Appropriate places already use `[[gnu::flatten]]` behind a `FLATTENED` macro, and I write all the tight loops to avoid needing to make non-inlinable function calls as well.  I get very good performance from even debug builds, as in good enough for realtime usage.

Unless there's a really good reason, trying to figure out how to build Synthizer with custom-built clang on Windows isn't something I particularly want to look into further at this time.  If you have either a POC benchmark that shows how it can be better or a source as to what the official Clang 10 build doesn't have that we might want I'm happy to revisit this, but cmake is fragile at the best of times and it builds both on my machine and in CI so I really need more justification as to why this is important before I allocate it time.

**Comment by Keithcat1** [comment]
I think the reason that -D_CMAKE_TOOLCHAIN_PREFIX=llvm might need to be specified is that I have the Mingw64 GCC compiler installed on my machine, and Cmake is using its tools instead of the LLVM ones.

**Comment by camlorn** [comment]
This is either the most requested feature or close to it, and is consequently now the next project.  We'll be stripping out the SIMD stuff and replacing it with a macro-based shim, which opens us up to compiling on MSVC directly.  This drops the Clang dependency for Windows.  Then I'll bash on gcc in wsl until I've got a Linux build, at which point I'll probably need one of the Linux people here to test it.  Though I will probably end up with a Linux vm soon for other reasons, so maybe I can just test it myself, we'll see.

I've been promising stuff after reverb for a while. We're now after reverb.  Unfortunately there's still no timeline.  This was my week off and I was hoping to get to it, but it's Friday already so probably not going to happen.  But at least there's a plan.

**Comment by camlorn** [comment]
We now compile with MSVC 2019 and GCC 9.3 or later.  I'm not sure if the GCC version can be pushed back or not yet.  Both under Windows and under WSL, the following simpler build procedure works:

```
mkdir build
cd build
cmake -G Ninja ..
ninja
```

if someone wants to verify whether or not we have audio on an actual Linux machine, that's the next step.  Not sure when I'll have the time to get a vm going.

It's probably not going to be possible to push the MSVC version back, unfortunately.  I think gcc can maybe be pushed to 8 but am not super enthusiastic about playing the "how do I downgrade the compiler" game.  Unless we need something older, I'll probably just create a spin-off issue for mac, integrate with CI, and close out this one.

**Comment by ndarilek** [comment]
I'll see about confirming Linux audio sometime next week and will report 
back. In theory I should be able to easily confirm that it outputs 
audio, but my WSL/Pulse setup can be pretty choppy. So I couldn't 
necessarily distinguish "outputs audio but there are Synthizer buffering 
issues" from "works just fine, my setup's just busted."


I can confirm macOS as well. Is there a particular reason you'd want to 
break it out into a separate issue? Do you not anticipate that you now 
support Clang, for instance? Because if you can compile on Clang, then 
it stands to reason that your macOS support is in the same uncertain 
place Linux support is, and you'd just need confirmation on both counts.

**Comment by camlorn** [comment]
Getting VMWare is on my backlog for other reasons anyway as I need a Linux VM in the near future for work, so if someone says "this is completely broken" there's a chance I can do something about it for Linux.  Getting someone else who's already set up with this to test is just the easiest short term path until that happens.

But mac literally needs another computer which I have neither the money nor the space for.  At this rate I need to move and get an office or something where I have enough space for some sort of nerdy tech lab, I guess.  So ideally someone who really wants it on that platform can step forward to at least diagnose/reproduce issues.  Ci for it is also going to take more effort, and I think there's something about compiling miniaudio as objective-c.

Broadly speaking though, either miniaudio works or it doesn't.  As things stand the rest of Synthizer is either pure math or using things from the C++ stdlib.  The strategy here overall is that if there's users on the platforms I don't have the bandwidth for, they'll report issues as necessary.  Manylinux1 and os x Python wheels will be available, which should get us some coverage as I know there's some adopters specifically waiting on those.

**Comment by ndarilek** [comment]
Right, I get all that. Let me be explicit: I do have a Mac. So as soon 
as things work with compilers other than the Microsoft toolchain, I'm 
happy to confirm whether or not it works for you. I just don't know if 
a) it compiles with Clang and b) it needs to in order to run on macOS, 
since I think macOS uses Clang by default.


FWIW, the Mac Mini is pretty small, and has been adequate enough for my 
dev needs. I get that money is also an issue, but in terms of space, I 
barely notice the space the Mini takes up.

**Comment by camlorn** [comment]
It works on gcc and clang already.  See my above comment.  I just can't confirm audio output.  At this point the compiler shouldn't be an issue anymore, but platform differences may still come up.

**Comment by ndarilek** [comment]
Got it, wasn't sure if GCC and Clang were functionally identical WRT 
your changes.


I'll poke both next week and report back here.

**Comment by Keithcat1** [comment]
I'm getting the following errors on WSL Ubuntu:
[1/3] Linking CXX executable property_write_bench
FAILED: property_write_bench 
: && /usr/bin/c++     CMakeFiles/property_write_bench.dir/benchmarks/property_write.cpp.o  -o property_write_bench  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `std::thread::thread<synthizer::Context::initContext()::{lambda()#2}, , void>(synthizer::Context::initContext()::{lambda()#2}&&)':
context.cpp:(.text+0x4275): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::Semaphore(int)':
context.cpp:(.text._ZN9SemaphoreC2Ei[_ZN9SemaphoreC5Ei]+0x48): undefined reference to `sem_init'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::~Semaphore()':
context.cpp:(.text._ZN9SemaphoreD2Ev[_ZN9SemaphoreD5Ev]+0x18): undefined reference to `sem_destroy'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::wait()':
context.cpp:(.text._ZN9Semaphore4waitEv[_ZN9Semaphore4waitEv]+0x18): undefined reference to `sem_wait'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::signal()':
context.cpp:(.text._ZN9Semaphore6signalEv[_ZN9Semaphore6signalEv]+0x18): undefined reference to `sem_post'
/usr/bin/ld: libsynthizer.a(memory.cpp.o): in function `std::thread::thread<void (&)(), , void>(void (&)())':
memory.cpp:(.text._ZNSt6threadC2IRFvvEJEvEEOT_DpOT0_[_ZNSt6threadC5IRFvvEJEvEEOT_DpOT0_]+0x33): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_create__posix':
single_file_libs.c:(.text+0x4524c): undefined reference to `pthread_attr_setstacksize'
/usr/bin/ld: single_file_libs.c:(.text+0x4533a): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_wait__posix':
single_file_libs.c:(.text+0x4538e): undefined reference to `pthread_join'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlopen':
single_file_libs.c:(.text+0x4600e): undefined reference to `dlopen'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlclose':
single_file_libs.c:(.text+0x46038): undefined reference to `dlclose'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlsym':
single_file_libs.c:(.text+0x46066): undefined reference to `dlsym'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_context_init_backend_apis__nix':
single_file_libs.c:(.text+0x55702): undefined reference to `pthread_create'
/usr/bin/ld: single_file_libs.c:(.text+0x55714): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
[2/3] Linking CXX executable test_noise
FAILED: test_noise 
: && /usr/bin/c++     CMakeFiles/test_noise.dir/test/noise.cpp.o  -o test_noise  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `std::thread::thread<synthizer::Context::initContext()::{lambda()#2}, , void>(synthizer::Context::initContext()::{lambda()#2}&&)':
context.cpp:(.text+0x4275): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::Semaphore(int)':
context.cpp:(.text._ZN9SemaphoreC2Ei[_ZN9SemaphoreC5Ei]+0x48): undefined reference to `sem_init'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::~Semaphore()':
context.cpp:(.text._ZN9SemaphoreD2Ev[_ZN9SemaphoreD5Ev]+0x18): undefined reference to `sem_destroy'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::wait()':
context.cpp:(.text._ZN9Semaphore4waitEv[_ZN9Semaphore4waitEv]+0x18): undefined reference to `sem_wait'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::signal()':
context.cpp:(.text._ZN9Semaphore6signalEv[_ZN9Semaphore6signalEv]+0x18): undefined reference to `sem_post'
/usr/bin/ld: libsynthizer.a(memory.cpp.o): in function `std::thread::thread<void (&)(), , void>(void (&)())':
memory.cpp:(.text._ZNSt6threadC2IRFvvEJEvEEOT_DpOT0_[_ZNSt6threadC5IRFvvEJEvEEOT_DpOT0_]+0x33): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_create__posix':
single_file_libs.c:(.text+0x4524c): undefined reference to `pthread_attr_setstacksize'
/usr/bin/ld: single_file_libs.c:(.text+0x4533a): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_wait__posix':
single_file_libs.c:(.text+0x4538e): undefined reference to `pthread_join'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlopen':
single_file_libs.c:(.text+0x4600e): undefined reference to `dlopen'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlclose':
single_file_libs.c:(.text+0x46038): undefined reference to `dlclose'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlsym':
single_file_libs.c:(.text+0x46066): undefined reference to `dlsym'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_context_init_backend_apis__nix':
single_file_libs.c:(.text+0x55702): undefined reference to `pthread_create'
/usr/bin/ld: single_file_libs.c:(.text+0x55714): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
[3/3] Linking CXX executable file_test
FAILED: file_test 
: && /usr/bin/c++     CMakeFiles/file_test.dir/file_test.cpp.o  -o file_test  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `std::thread::thread<synthizer::Context::initContext()::{lambda()#2}, , void>(synthizer::Context::initContext()::{lambda()#2}&&)':
context.cpp:(.text+0x4275): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::Semaphore(int)':
context.cpp:(.text._ZN9SemaphoreC2Ei[_ZN9SemaphoreC5Ei]+0x48): undefined reference to `sem_init'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::~Semaphore()':
context.cpp:(.text._ZN9SemaphoreD2Ev[_ZN9SemaphoreD5Ev]+0x18): undefined reference to `sem_destroy'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::wait()':
context.cpp:(.text._ZN9Semaphore4waitEv[_ZN9Semaphore4waitEv]+0x18): undefined reference to `sem_wait'
/usr/bin/ld: libsynthizer.a(context.cpp.o): in function `Semaphore::signal()':
context.cpp:(.text._ZN9Semaphore6signalEv[_ZN9Semaphore6signalEv]+0x18): undefined reference to `sem_post'
/usr/bin/ld: libsynthizer.a(memory.cpp.o): in function `std::thread::thread<void (&)(), , void>(void (&)())':
memory.cpp:(.text._ZNSt6threadC2IRFvvEJEvEEOT_DpOT0_[_ZNSt6threadC5IRFvvEJEvEEOT_DpOT0_]+0x33): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_create__posix':
single_file_libs.c:(.text+0x4524c): undefined reference to `pthread_attr_setstacksize'
/usr/bin/ld: single_file_libs.c:(.text+0x4533a): undefined reference to `pthread_create'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_thread_wait__posix':
single_file_libs.c:(.text+0x4538e): undefined reference to `pthread_join'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlopen':
single_file_libs.c:(.text+0x4600e): undefined reference to `dlopen'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlclose':
single_file_libs.c:(.text+0x46038): undefined reference to `dlclose'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_dlsym':
single_file_libs.c:(.text+0x46066): undefined reference to `dlsym'
/usr/bin/ld: libsynthizer.a(single_file_libs.c.o): in function `ma_context_init_backend_apis__nix':
single_file_libs.c:(.text+0x55702): undefined reference to `pthread_create'
/usr/bin/ld: single_file_libs.c:(.text+0x55714): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

**Comment by ndarilek** [comment]
See if forcing a link against pthreads and installing something like 
libpthreads-dev gets that working. I don't know cmake well enough to 
know the incantations, but it shouldn't be hard to find whatever's 
necessary to add -lpthreads or whatever.


If that works, for bonus points, figure out how to conditionally link 
that for Linux only. :) Sorry I can't be more helpful--I don't have time 
to hack on this until next week, but at least until then I can offer 
pointers.

**Comment by Keithcat1** [comment]
OK, I've got it down to:
[1/3] Linking CXX executable property_write_bench
FAILED: property_write_bench 
: && /usr/bin/c++  -pthread  -pthread -Wl,--no-as-needed -ldl CMakeFiles/property_write_bench.dir/benchmarks/property_write.cpp.o  -o property_write_bench  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
collect2: error: ld returned 1 exit status
[2/3] Linking CXX executable test_noise
FAILED: test_noise 
: && /usr/bin/c++  -pthread  -pthread -Wl,--no-as-needed -ldl CMakeFiles/test_noise.dir/test/noise.cpp.o  -o test_noise  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
collect2: error: ld returned 1 exit status
[3/3] Linking CXX executable file_test
FAILED: file_test 
: && /usr/bin/c++  -pthread  -pthread -Wl,--no-as-needed -ldl CMakeFiles/file_test.dir/file_test.cpp.o  -o file_test  libsynthizer.a && :
/usr/bin/ld: libsynthizer.a(stereo_panner.cpp.o): in function `synthizer::StereoPanner::getLane(unsigned int)':
stereo_panner.cpp:(.text+0x190): undefined reference to `synthizer::StereoPanner::LANES'
/usr/bin/ld: libsynthizer.a(hrtf.cpp.o): in function `synthizer::HrtfPanner::getLane(unsigned int)':
hrtf.cpp:(.text+0xe92): undefined reference to `synthizer::HrtfPanner::CHANNELS'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

I got rid of the other errors by doing:
export CFLAGS=-pthread
export CXXFLAGS=-pthread
export LDFLAGS="-pthread -Wl,--no-as-needed -ldl"

**Comment by camlorn** [comment]
@ndarilek 
the feature here is that you no longer need clang; I've supported that from day one.

@Keithcat1 
That looks like an old commit. Are you sure you've pulled latest master?  I remember explicitly fixing some of those errors for gcc.

CMake should also be linking threads conditionally as well as dlopen on all platforms.  They provide that logic, I wired it in, it worked for me.

What's the compiler and version?  maybe clang on Linux explodes after all?

**Comment by camlorn** [comment]
Ok, apparently clang on Linux is in fact not working even though clang on Windows does. I'm investigating. It's almost certainly a CMake issue.

**Comment by camlorn** [comment]
@Keithcat1 
Compile like this:

```
CC=clang CXX=clang++ cmake -G Ninja ..
```

If that works, it's not a Synthizer bug.

**Comment by camlorn** [comment]
Ok, sorry everyone.  I forgot to commit some changes that made this work and they've just been sitting there unpushed since my initial comments.  Hopefully things are better now.

My comment above still applies, though: if trying to build with clang on Linux when it's not already the default compiler for your system, use `CC=clang CXX=clang++`

**Comment by ndarilek** [comment]
Odd, because I'm pretty sure I forced a build with Clang under Linux, 
and it failed due to what looked like language syntax issues. I'd 
assumed MSVC-specific extensions were used, or that Clang needed a 
switch to enable a newer C++ standard.


Anyway, going to work on my own projects today, but I'll earmark some 
time to attempt running on Linux virtualized or macOS on Thursday or so.

**Comment by ndarilek** [comment]
Never mind, saw later messages in this thread. Apologies.

**Comment by Keithcat1** [comment]
Can you find a way to only export the C API if a shared library is being built?
I'm ending up with executables that are exporting symbols.
C:\py\git\synthizer\build>dumpbin /exports file_test.exe
Microsoft (R) COFF/PE Dumper Version 14.27.29112.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file file_test.exe

File Type: EXECUTABLE IMAGE

  Section contains the following exports for file_test.exe

    00000000 characteristics
    FFFFFFFF time date stamp
        0.00 version
           1 ordinal base
          30 number of functions
          30 number of names

    ordinal hint RVA      name

          1    0 00008756 syz_bufferGetChannels = @ILT+30545(syz_bufferGetChannels)
          2    1 0000E8BD syz_bufferGetLengthInSamples = @ILT+55480(syz_bufferGetLengthInSamples)
          3    2 0000917E syz_bufferGetLengthInSeconds = @ILT+33145(syz_bufferGetLengthInSeconds)
          4    3 0000F2D6 syz_configureLoggingBackend = @ILT+58065(syz_configureLoggingBackend)
          5    4 00001E92 syz_createBufferFromStream = @ILT+3725(syz_createBufferFromStream)
          6    5 0000F0FB syz_createBufferGenerator = @ILT+57590(syz_createBufferGenerator)
          7    6 0000650A syz_createContext = @ILT+21765(syz_createContext)
          8    7 00003DBE syz_createGlobalFdnReverb = @ILT+11705(syz_createGlobalFdnReverb)
          9    8 00006D11 syz_createPannedSource = @ILT+23820(syz_createPannedSource)
         10    9 00005D5D syz_createSource3D = @ILT+19800(syz_createSource3D)
         11    A 0000EFDE syz_getD = @ILT+57305(syz_getD)
         12    B 000087F1 syz_getD3 = @ILT+30700(syz_getD3)
         13    C 000019B5 syz_getD6 = @ILT+2480(syz_getD6)
         14    D 000039F4 syz_getI = @ILT+10735(syz_getI)
         15    E 00007DE2 syz_getLastErrorCode = @ILT+28125(syz_getLastErrorCode)
         16    F 0000A411 syz_getLastErrorMessage = @ILT+37900(syz_getLastErrorMessage)
         17   10 0000EB51 syz_getO = @ILT+56140(syz_getO)
         18   11 00004D2C syz_handleFree = @ILT+15655(syz_handleFree)
         19   12 00002D6A syz_initialize = @ILT+7525(syz_initialize)
         20   13 00002BD0 syz_routingConfigRoute = @ILT+7115(syz_routingConfigRoute)
         21   14 0000E58E syz_routingRemoveRoute = @ILT+54665(syz_routingRemoveRoute)
         22   15 0000C02C syz_setD = @ILT+45095(syz_setD)
         23   16 00009CDC syz_setD3 = @ILT+36055(syz_setD3)
         24   17 0000E3F9 syz_setD6 = @ILT+54260(syz_setD6)
         25   18 00003080 syz_setI = @ILT+8315(syz_setI)
         26   19 00001037 syz_setLogLevel = @ILT+50(syz_setLogLevel)
         27   1A 0000BEC9 syz_setO = @ILT+44740(syz_setO)
         28   1B 0000DA3A syz_shutdown = @ILT+51765(syz_shutdown)
         29   1C 00003602 syz_sourceAddGenerator = @ILT+9725(syz_sourceAddGenerator)
         30   1D 000026D5 syz_sourceRemoveGenerator = @ILT+5840(syz_sourceRemoveGenerator)

  Summary

        1000 .00cfg
      259000 .data
        2000 .idata
       25000 .pdata
       A3000 .rdata
        8000 .reloc
        1000 .rsrc
      29C000 .text
       10000 .tls
        1000 _RDATA

**Comment by Keithcat1** [comment]
Also: I'm using GCC 10.2 (msys64) and src/memory.cpp is taking at least half an hour to compile (in release mode only).
I don't know how long it actually takes because it wasn't done after I last checked.
my build procedure was pretty normal, cmake -G Ninja -DCMAKE_BUILD_TYPE=release .. followed by ninja

**Comment by camlorn** [comment]
@Keithcat1 
yeah, the symbol thing can be fixed.

The memory.cpp thing I have no idea on.  The library builds in 2 or 3 minutes even on release builds.  I've seen it work on all 3 major compilers.  I'm not willing to install msys to try to see if I can duplicate it that way because every time I try to get msys to play nice with everything else on Windows I can't manage it.  I would be literally unable to make a C++ file that takes half an hour to compile on any machine if I tried.  If other people start having this problem I'll give it more priority and if I can figure out how to get multiple versions of gcc going on WSL or something and I can duplicate it I'll try to fix it.  But you're saying that WSL GCC 9.3 works, Msys GCC 10 doesn't, and not only does Msys GCC not work, it doesn't work in a way where we'd have to file a bug against GCC for deciding to hang for half an hour rather than error, so at this time I can't justify going down that rabbit hole.  Until more evidence proves otherwise, I'm going to have to say that it's probably something about your machine, that gcc has always treated Windows as a very much second class citizen, and that if you send a patch I'll be happy to review it.

**Comment by Keithcat1** [comment]
Maybe I should try a different GCC compiler then. Where did you get your windows version of GCC from because it seems to do a better job than what I've got. The only other compiler I've got is GCC 8.1, and I know Synthizer doesn't support that.

**Comment by camlorn** [comment]
I don't get gcc for Windows because gcc on Windows is unnecessary for most use cases and, every time I have seriously tried to use it, I've run up against the thing where they don't support enough MSVC extensions to parse the official headers from Microsoft.  Being as Clang is now officially supported by Microsoft and 99% gcc compatible and gcc isn't supported by Microsoft and is prone to breaking when you try to use it, I don't intend to ever bother.  But in this case, it should work, and the fact that it doesn't is almost certainly not a Synthizer problem.

This has shown that I need to be really clear, however.  I intend to support the following combinations.  If your combination isn't on this list I am willing to add it with justification.  I'm only one person, this needs to be scoped, and these are the most valuable targets as I see it:

- For Windows, Clang 9.0 or later, MSVC 2019. Could be convinced to try for older MSVC, but older MSVC is really bad at being standards compliant.
- For Linux, clang 9.0 or later, gcc 9 or later. I might be able to push us back to older versions of these compilers.
- For Apple, whatever the current version of Xcode is.

And also, if Synthizer is crashing or freezing your compiler, please start under the assumption that it's not a Synthizer bug until proven otherwise.

I regret having to play hardball a bit here, but this issue and the one related to Python on Linux are slowly turning into asking me to support every combination and deployment strategy out there.  There needs to be success criteria.  Or more knowledgeable C++ people.  Whichever works for me.

**Comment by camlorn** [comment]
Well, okay. So fun times.  Gcc 10 is broken on my Linux VM as well now.  It looks like it can't handle large ConcurrentSlab capacities? Looks like this will get fun.

**Comment by camlorn** [comment]
@ndarilek 
As of now, I've confirmed audio output on Linux against Alsa.  I can't confirm Pulse.  Orca works, but I don't know if that's Pulse or not.  I suspect that Pulse not working is a combination of me being on Vmware and Synthizer maybe not being so great w.r.t. threads the way Pulse is trying to schedule.  If you have actual hardware with Pulse installed, it'd be worth a  test at some point to know which one it is.  Otherwise, I'll re-check this once I've done some of the planned work around optimizing the command queues, which just became higher priority.  VMWare is a challenging environment for audio, though, and a lot of the stuff I read about Pulse is "well it doesn't work sometimes, sorry", so I'm not sure what to make of it.

Worst case is someone with actual Pulse experience needs to take it over.  Hoping that doesn't happen, but properly configuring and debugging audio apparently involves upwards of 5 config files.

Also to get ahead of "why not just turn off Pulse in the code?"  because Pulse emulates Alsa in a way that's worse than the legacy Windows APIs that now have almost unusable latency, so it sorta ends up not mattering.

@Keithcat1 
Congratulations on finding the first GCC bug in my C++ career.  I'm thinking regression, but don't have the bandwidth to investigate and properly report.  In either case, it should now be fixed.

It looks like something involved with their handling of constexpr or something is quadratic or worse: the issue was that `ConcurrentSlab` wanted to initialize a bunch of really big arrays with nontrivial constructors at compile time.  I've moved them to runtime, so there's a little bit of a performance cost there now, but better than not compiling.

**Comment by camlorn** [comment]
Ok, Linux Pulseaudio appears to be working, at the cost of increased property read performance issues.  My philosophy on that is that the manual makes it clear that property reads are slow and for debugging.  The short version of a long conversation is that stable audio on Linux requires 3 or 4 settings, most notably `SCHED_FIFO` if you can get it, and Miniaudio knows how to set all that up.  My guess is that the pulse server was able to preempt Synthizer's audio thread previously.

Tracking property read performance in #32.  There's probably a different set of trade-offs that are eventually consistent that would let us do better, but writing up that design requires writing an entire developer's manual for the code first.

**Comment by camlorn** [comment]
Got some extra money for hardware.  Anyone have thoughts on whether or not developing against the new M1 macs is going to be sufficient?  At some point I need an ARM device for optimizing against, and it would be nice to knock out two birds with one stone (GitHub has x86 Macs  in workflow, and we're going to be moving to workflow for other reasons, which is an entire separate issue. So this doesn't mean no Mac wheels for Python users).

@ndarilek 
You have some hardware you like for surround sound, I think.  Remind me what that is?  Also any recs on an Android tablet?  Given that I've got some money I can devote to this, thinking about setting up to be able to support the phones while I'm at it.

**Comment by camlorn** [comment]
Linux is now complete as far as I'm concerned.  As of 0.7.6 you can `pip install synthizer` etc. and I've had multiple confirmations that audio output works fine.  Bumping this to the 1.0 milestone and assuming that it's basically the mac issue at this point.

**Comment by ndarilek** [comment]
For surround, I got the 45-inch version of this. Note that it goes down to 32-inch:

https://www.amazon.com/VIZIO-SB3651-E6B-Soundbar-Speaker-Renewed/dp/B07963TVWG/ref=sr_1_3?dchild=1&keywords=vizio+5.1+surround+bar&link_code=qs&qid=1606579149&sourceid=Mozilla-search&sr=8-3&tag=mozilla-20

While I do have it set up and generally like it, I will offer the disclaimer that it was a pain and a half to set up and I don't know if I recommend it anymore. Once you get it on your network, it is controllable via Python library and CLI. But I had a hard time using both Android and iOS apps. I think Android has some sort of NFC setup ability, but I could never find the embedded tag no matter how hard I tried. Maybe it's easier with sighted help, but that isn't necessarily easy to come by these days.

I'm not looking to replace this thing. For a 45-inch soundbar, that I paid $200 or so for, the sound isn't too bad for my space. The sub is flat, and sits on its side under my desk. The soundbar takes up about a couple inches of thickness on the shelf in front of my monitors. The rears are screwed into the wall above my bed, and are pretty small. I have it set up the way I like it, and hope I don't have to do it again anytime soon. :) If I do, I'll bribe someone to help.

**Comment by camlorn** [comment]
This got rolled back in 0.7.7 while also demonstrating I may need a better branching structure to do emergency point releases.  @chrisnorman7 has managed to prove that the performance regressions in #32 are in fact real world, so this is now blocked on me finishing that refactor and then figuring out if there's anything else that got introduced as part of this work which may cause issues.  Think it's probably #32 though, because what changed here is that the command queue doesn't wake up immediately and we have to block on it for object creation.

Apologies.  I will correct this as soon as possible, but there's no other way to do Linux but through.

**Comment by chrisnorman7** [comment]
Still a great library. I’m happy to wait.

**Comment by camlorn** [comment]
In the interest of triaging appropriately, Linux is back at this point unless someone tells me otherwise.  We'll be adding that to CI in the near future, which will hopefully remove the "unless someone tells me otherwise" clarification.  Until then, don't hesitate to tell me otherwise.

**Comment by Keithcat1** [comment]
You broke GCC again:
../test/verify_properties.cpp:108:6: error: ‘isinf’ was not declared in this scope; did you mean ‘std::isinf’?
There's a bunch of "already defined" warnings as well. It looks like a simple-ish fix.

**Comment by Keithcat1** [comment]
OK, you didn't break GCC. You just broke the test_verify_properties test, which in turn broke cmake -G Ninja .. && ninja all
I can still build the synthizer library its self and it just throws some warnings at me.

**Comment by camlorn** [comment]
I'll look into getting quick GCC fixes. Sounds like we need more pragmas.

The ultimate plan is GitHub workflow for Linux and -Wall on all platforms, but we aren't there yet.

**Comment by camlorn** [comment]
Ok, was all quick fixes.  I'll push it out in the next release being as the library currently builds.

**Comment by ahicks92** [comment]
We now have Linux CI configured to block releases.  This is against Ubuntu 20.  If anyone wants to nominate some sort of official Linux compiler support policy I'm open to it, as long as there's more thought behind it than "I like these numbers".

**Comment by ahicks92** [comment]
Mac support is done.  I'll close this issue after 0.9, once we see a successful pip install.

I want iOS and Android but will track those separately after 1.0, contingent on someone saying they want to spend a couple man-weeks on it: those require bindings to be useful and, particularly in the case of Android, writing said bindings isn't trivial.  If no one is willing to become the long-term maintainer for those we won't get them.

The OSX CI is against OSX 10.  I compiled it on 11, but 11 is a private beta at the moment.  Note also that I won't be able to support more than the latest version of OSX that GitHub actions supports.  They multiply the minutes used by OSX by 10, so our one minute OSX job counts for 10 against the usage limit.  I'll do my best to support older OSX if there proves to be a need, but there's very little I can do on the GitHub actions front or the personal hardware front.

Early adopters who already know how to build the repo can build the -0.9-dev branch to try it.  Note that the 0.9-dev branch is in flux: things may or may not work, especially in the Python bindings, which I've intentionally let run behind as I continue refactoring and improving the C API.

**Comment by chrisnorman7** [comment]
Good work. I'll have a look at the branch Tuesday or Wednesday probably.

I'd love to offer to be the Android person, but I don't have the skills,
and honestly, with the lack of a headphone port on most modern phones, I'm
not even sure how much use something like Synthizer would get.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
Eventually I want to take some of my audiogaming projects to the phones but that's literally years away.  The C library is honestly pretty close to compiling, but you can't test any of it without a full app.  Also, without Swift and Java bindings Saying that it supports the phones is one of those technically true but not really statements.  For blind people/others interested in HRTF, low-latency bluetooth is now a thing; for everyone else, there's a lot of value in being able to just use the same lib on all 5 platforms (and remember that surround sound support is before 1.0 as well).  But for myself at some point I want to use Synthizer for the stuff I built it for rather than this being my side project life forever.

**Comment by chrisnorman7** [comment]
Fair enough, and are you hoping to distribute your stuff to mobile too?

If you need a full app making, I can knock you something together in
Flutter, if you let me know what you need. I'm not sure what that would
entail, but it might be quicker for me to do it, being as I already know
Dart and the Flutter framework.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
Down the road, yes.  I don't know what, when, or how much.  Mostly I'll be adopting Rust which is kind of trivially cross-platform except for the UI, so.  Unfortunately Flutter isn't going to do it from the perspective of being able to claim support with a straight face.  I might go that route for my projects, but if I seriously want to claim to support mobile it does mean Swift and Java and I suspect that I need more first-party tooling for debuggability and things like that for development.  I appreciate the offer though.

**Comment by ahicks92** [comment]
Mac is on Pypi and done. Closing this out per my above comments.
